pyramid/tier_fs.go

```diff
@@ -253,12 +268,18 @@ func (tfs *TierFS) readFromBlockStorage(fileRef localFileRef) (*os.File, error)
 		if err != nil {
 			return nil, fmt.Errorf("copying date to file: %w", err)
 		}
+		downloadHistograms.WithLabelValues(tfs.fsName).Observe(float64(written))
 
 		if err = writer.Close(); err != nil {
-			err = fmt.Errorf("writer close: %w", err)
+			return nil, fmt.Errorf("writer close: %w", err)
 		}
-		downloadHistograms.WithLabelValues(tfs.fsName).Observe(float64(written))
-		return nil, err
+
+		// copy from temp path to actual path
+		if err = tfs.syncDir.renameFile(tmpFullPath, fileRef.fullPath); err != nil {
+			return nil, fmt.Errorf("rename temp file: %w", err)
+		}
+
+		return nil, nil
```

Comment:
```suggestion
		return fh, nil
```

Reply:
Again, this is really a race.  On line 289 you *again* try to open the file.  But the file might already have been removed.  This is rare, but easy enough to fix: keep the return value from `keyLock.Compute` and return it here if possible.

Reply:
I agree that it's a better habit to return the file-handler and I'll fix it.
However notice it's not a race. The file wasn't stored in the eviction cache yet, it happens in the `openFile` func.
So file wasn't stored in the eviction -> can't be evicted -> file is always there.

Reply:
Actually, after applying the change, it is wrong.
`Compute` only calls the function once for concurrent calls, so returning the same file-handler isn't right.

Reply:
Yikes!  True...  So the way to do this would be to keep an external `var fh *os.File`, and set that here (and still `return nil, nil`), then use that `fh` if not `nil` after this `Compute` call.  Which is unpleasant.  (oh, and I forgot to rewind the file before returning, of course).
I still don't like this double-open (you open a file twice when you download it), but I do not see a race to cause trouble here - just some inefficiency.